### PR TITLE
Fix Use Counters at the Service Level

### PR DESCRIPTION
**Pull request by chutten**
[bug 1412382](https://bugzilla.mozilla.org/show_bug.cgi?id=1412382)

Use counters in the database are wrong, and have been since [bug 1204994](https://bugzilla.mozilla.org/show_bug.cgi?id=1204994) changed
_extract_main_histograms. It only counted "False" values in pings that had at
least one "True" value (because if there were only "False" values, we didn't
send a use counter in that ping).

This fixes this by subbing in the False value from
(TOP_LEVEL_)CONTENT_DOCUMENTS_DESTROYED, which has the correct number.

This results in an interesting side-effect that use counters that don't exist
will get valid responses from the service. This is because the service can't
tell the difference between a use counter that doesn't exist and one that just
didn't happen to have a single 'True' value in that row in that table.

This is beneficial for testing, so that I don't have to manipulate
histograms_template or generate_payload in dataset.py to be able to support
probes that aren't in every ping. The math for expected_count in test_db
would only get worse.

Instead I can ask for a use counter that doesn't exist and ensure that it
reports the correct number of False values and the correct count.

### CODE REVIEW

File: mozaggregator/aggregator.py [hunk 1]
```diff
@@ -170,48 +170,7 @@ def _extract_main_histograms(state, histograms, process_type):
     if not isinstance(histograms, dict):
         return
 
-    # Deal with USE_COUNTER2_ histograms, see Bug 1204994
```
Comment (fbertsch) [inline mozaggregator/aggregator.py]: All this red: Beautiful. Absolutely beautiful.

File: mozaggregator/service.py [hunk 2]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
         altered_dimensions['child'] = reverse_map.get(altered_dimensions['child'], altered_dimensions['child'])
 
     # Fetch metrics
-    result = execute_query("select * from batched_get_metric(%s, %s, %s, %s, %s, %s)", (prefix, channel, version, dates, json.dumps(dimensions), json.dumps(altered_dimensions)))
+    result = None
```
Comment (fbertsch) [inline mozaggregator/service.py]: No need to include this, `result` will be in scope after the `if/else`.

File: mozaggregator/service.py [hunk 3]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
...
+                t1.histogram[3],
+                t1.histogram[4],
+                t1.histogram[5]]
+            END,
+            t1.histogram,
+            t2.histogram
+          from batched_get_metric(%s, %s, %s, %s, %s, %s) t1
+          right outer join batched_get_metric(%s, %s, %s, %s, %s, %s) t2
+          on t1.date = t2.date
+          """, (prefix, channel, version, dates, json.dumps(dimensions), json.dumps(altered_dimensions),
```
Comment (fbertsch) [inline mozaggregator/service.py]: Instead of passing the same values many times, a dictionary with named values can be passed instead, so the `prefix`, `channel`, `version`, and `dates` can be reused: http://initd.org/psycopg/docs/usage.html#query-parameters

File: mozaggregator/service.py [hunk 4]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
...
+                t2.histogram[array_length(t2.histogram, 1) - 1] - t1.histogram[2] - t1.histogram[3],
+                t1.histogram[2],
+                t1.histogram[3],
+                t1.histogram[4],
+                t1.histogram[5]]
+            END,
+            t1.histogram,
+            t2.histogram
+          from batched_get_metric(%s, %s, %s, %s, %s, %s) t1
+          right outer join batched_get_metric(%s, %s, %s, %s, %s, %s) t2
```
Comment (fbertsch) [inline mozaggregator/service.py]: I think this should be a `full outer join`, to capture the possibility of a use counter being reported, but not a `denominator`. We'll then have to deal with that possibility above.

File: mozaggregator/service.py [hunk 5]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
...
+            coalesce(t1.label, ''),
+            CASE
+              WHEN t1.histogram IS NULL THEN ARRAY[
+                t2.histogram[array_length(t2.histogram, 1) - 1],
+                0,
+                0,
+                0,
+                t2.histogram[array_length(t2.histogram, 1)]]
+              ELSE ARRAY[
+                t2.histogram[array_length(t2.histogram, 1) - 1] - t1.histogram[2] - t1.histogram[3],
```
Comment (fbertsch) [inline mozaggregator/service.py]: Why subtract both `histogram[2]` and `histogram[3]`?

File: mozaggregator/service.py [hunk 6]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
...
+        if metric.endswith("_DOCUMENT"):
+            denominator = "CONTENT_DOCUMENTS_DESTROYED"
+        denominator = "{}_{}".format(COUNT_HISTOGRAM_PREFIX, denominator)
+        denominator_dimensions = deepcopy(dimensions)
+        denominator_dimensions["metric"] = denominator
+        denominator_new_dimensions = deepcopy(altered_dimensions)
+        denominator_new_dimensions["metric"] = denominator
+        result = execute_query("""
+          select t2.date,
+            coalesce(t1.label, ''),
```
Comment (fbertsch) [inline mozaggregator/service.py]: `batched_get_metric` just returns the label itself here (`NULL` or not), any particular reason to `coalesce` to an empty string?

File: mozaggregator/service.py [hunk 7]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
...
+        # values and False values supplied by *CONTENT_DOCUMENTS_DESTROYED.
+        denominator = "TOP_LEVEL_CONTENT_DOCUMENTS_DESTROYED"
+        if metric.endswith("_DOCUMENT"):
+            denominator = "CONTENT_DOCUMENTS_DESTROYED"
+        denominator = "{}_{}".format(COUNT_HISTOGRAM_PREFIX, denominator)
+        denominator_dimensions = deepcopy(dimensions)
+        denominator_dimensions["metric"] = denominator
+        denominator_new_dimensions = deepcopy(altered_dimensions)
+        denominator_new_dimensions["metric"] = denominator
+        result = execute_query("""
```
Comment (fbertsch) [inline mozaggregator/service.py]: Please move this query to the `sql.py` file.

File: tests/test_db.py [hunk 8]
```diff
@@ -151,6 +151,30 @@ def test_build_id_metrics():
                 test_keyed_histogram("build_id", channel, version, template_build_id, metric, histograms, histogram_expected_count)
 
 
+def test_absent_use_counter():
```
Comment (fbertsch) [inline tests/test_db.py]: Seems to me we should also test a non-empty use counter.

File: mozaggregator/service.py [hunk 9]
```diff
@@ -322,7 +322,44 @@ def get_dates_metrics(prefix, channel):
...
+                0,
+                t2.histogram[array_length(t2.histogram, 1)]]
+              ELSE ARRAY[
+                t2.histogram[array_length(t2.histogram, 1) - 1] - t1.histogram[2] - t1.histogram[3],
+                t1.histogram[2],
+                t1.histogram[3],
+                t1.histogram[4],
+                t1.histogram[5]]
+            END,
+            t1.histogram,
```
Comment (fbertsch) [inline mozaggregator/service.py]: These seem to be unused.

File: mozaggregator/service.py (hunk 2 above)
Comment (chutten) [reply mozaggregator/service.py]: Is that a python thing?

File: mozaggregator/service.py (hunk 7 above)
Comment (chutten) [reply mozaggregator/service.py]: Makes sense. I could do the sql param multiplexing at the SQL level that way.

To be clear, you're after something that looks like
```python
result = execute_query("select * from batched_get_use_counter(<param_list>)", <param list>)
```

And then all the fancy SQL shenanigans will live with the other SQL, right?

File: mozaggregator/service.py (hunk 6 above)
Comment (chutten) [reply mozaggregator/service.py]: In case label is absent (right join)

File: mozaggregator/service.py (hunk 5 above)
Comment (chutten) [reply mozaggregator/service.py]: The denominator is the total number of documents. `histogram[2]` is the number of documents for which it was true. `histogram[3]` is the number of documents for which something weird happened.

File: mozaggregator/service.py (hunk 9 above)
Comment (chutten) [reply mozaggregator/service.py]: Whoops

File: mozaggregator/service.py (hunk 4 above)
Comment (chutten) [reply mozaggregator/service.py]: Data is Noisy, so therefore it is _possible_ that this might happen... but I'm not sure what we want to return in that case. This would mean that, for a given set of dimensions, no documents were seen... but those documents used certain features.

I'm not sure we care to support this case.

File: mozaggregator/service.py (hunk 3 above)
Comment (chutten) [reply mozaggregator/service.py]: Nifty! But since I have to rewrite it as its own procedure in sql.py, I may not get a chance to use it.

File: tests/test_db.py (hunk 8 above)
Comment (chutten) [reply tests/test_db.py]: We already do. There are at least... three? `USE_COUNTER2` histograms in the dataset that are already tested.

**Comment by chutten** [comment]
@fbertsch Ready for another look.

**Comment by fbertsch** [comment]
@chutten I believe we want to support the "Has use counter, but not denominator". The issue with the current approach is that data will be totally unavailable in the service, since the `LEFT JOIN` won't return anything. I think it's pretty straightforward to add it now, with the `0` bucket value just being `0`. The client will need to figure out what to do with that information.

**Comment by chutten** [comment]
@fbertsch Fair enough. I've changed it to a `full outer join` and put in the third case.

**Comment by fbertsch** [comment]
@chutten one last change: `from db import _preparedb`, and:

```
def create_pool():
     global pool
     if pool is None:
+        _preparedb()
```

This will ensure the service refreshes the available procedures when we restart the service.

**Comment by chutten** [comment]
@fbertsch I appended the import requirement to the existing import line and added the requested code.
